lib/iris/fileformats/netcdf.py

```diff
@@ -1818,29 +2051,149 @@ def _get_coord_variable_name(self, cube, coord):
         else:
             name = coord.standard_name or coord.long_name
             if not name or set(name).intersection(string.whitespace):
-                # NB don't know how to fix this if it is not a Coord or similar
-                assert isinstance(coord, iris.coords.Coord)
-                # Auto-generate name based on associated dimensions.
-                name = ""
-                for dim in cube.coord_dims(coord):
-                    name += "dim{}".format(dim)
-                # Handle scalar coordinate (dims == ()).
-                if not name:
-                    name = "unknown_scalar"
+                # We need to invent a name, based on its associated dimensions.
+                if cube.coords(coord):
+                    # It is a regular cube coordinate.
+                    # Auto-generate a name based on the dims.
+                    name = ""
+                    for dim in cube.coord_dims(coord):
+                        name += "dim{}".format(dim)
+                    # Handle scalar coordinate (dims == ()).
+                    if not name:
+                        name = "unknown_scalar"
+                else:
+                    # Not a cube coord, so must be a connectivity or
+                    # element-coordinate of the mesh.
+                    # Name it for it's first dim, i.e. mesh-dim of its location.
+                    from iris.experimental.ugrid import Connectivity
+
+                    if isinstance(coord, Connectivity):
+                        location = coord.cf_role.split("_")[0]
+                    else:
+                        # Must be a mesh-element coordinate.
+                        location = None
+                        for test_location in MESH_LOCATIONS:
+                            include_key = f"include_{test_location}s"
+                            if coord in cube.mesh.coords({include_key: True}):
```

Comment:
`**{include_key: True}` ?

Reply:
As in e.g. `mesh.coords(include_nodes=True)`

I do wish that Mesh had an API giving access to these things programmatically, without needing to construct a keyword name from a location string.
So in this case, I think we should be able to say e.g. `mesh.coords(location='edge')`
And likewise in the case of connectivities, it would need something like `mesh.connectivities(from='face', to='node', axis='x')`
(but where  keys are variables, not code literals which we can already do)
Whereas I am currently using [keyword construction for that too](https://github.com/SciTools/iris/blob/813bfce4c1c1e4c763f9f9ffab9dac286ceccd77/lib/iris/fileformats/netcdf.py#L1786-L1787).

Reply:
@pp-mo You still need to change this, right? i.e., you can't just pass a dictionary here

Reply:
Ok I see what it was.
`test_location_coord_names` was supposed to cover this, but we can't actually have mesh location coords with no standard_name, so it is unreachable.

Removed this code + added some notes in case we relax this limitation later.

Reply:
Cool, thanks :+1: